presto-geospatial-toolkit/src/test/java/com/facebook/presto/geospatial/rtree/TestFlatbush.java

```diff
@@ -0,0 +1,281 @@
+import java.util.Comparator;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Collectors;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+public class TestFlatbush
+{
+    private static final Rectangle everything = new Rectangle(
+            Double.NEGATIVE_INFINITY,
+            Double.NEGATIVE_INFINITY,
+            Double.POSITIVE_INFINITY,
+            Double.POSITIVE_INFINITY);
+
+    private static final Comparator<Rectangle> rectangleComparator = Comparator
+            .comparing(Rectangle::getXMin)
+            .thenComparing(Rectangle::getYMin)
+            .thenComparing(Rectangle::getXMax)
+            .thenComparing(Rectangle::getYMax);
+
+    //  2 intersecting polygons: A and B
+    private static final OGCGeometry POLYGON_A = OGCGeometry.fromText("POLYGON ((0 0, -0.5 2.5, 0 5, 2.5 5.5, 5 5, 5.5 2.5, 5 0, 2.5 -0.5, 0 0))");
+    private static final OGCGeometry POLYGON_B = OGCGeometry.fromText("POLYGON ((4 4, 3.5 7, 4 10, 7 10.5, 10 10, 10.5 7, 10 4, 7 3.5, 4 4))");
+
+    // A set of points: X in A, Y in A and B, Z in B, W outside of A and B
+    private static final OGCGeometry POINT_X = TestUtils.makePoint(1, 1);
+    private static final OGCGeometry POINT_Y = TestUtils.makePoint(4.5, 4.5);
+    private static final OGCGeometry POINT_Z = TestUtils.makePoint(6, 6);
+    private static final OGCGeometry POINT_W = TestUtils.makePoint(20, 20);
+
+    @Test
+    public void testEmptyFlatbush()
+    {
+        Rectangle[] items = {};
+        Flatbush rtree = new Flatbush<Rectangle>(items);
+        assertEquals(rtree.findIntersections(everything).size(), 0);
+    }
+
+    @Test
+    public void testSingletonFlatbush()
+    {
+        List<Rectangle> items = new ArrayList();
+        items.add(new Rectangle(0, 0, 1, 1));
+        Flatbush rtree = new Flatbush<Rectangle>(items.toArray(new Rectangle[] {}));
+        assertEquals(rtree.findIntersections(everything), items);
+        Rectangle hit = new Rectangle(1, 1, 2, 2);
+        assertEquals(rtree.findIntersections(hit), items);
+        Rectangle miss = new Rectangle(-1, -1, -0.1, -0.1);
+        assertEquals(rtree.findIntersections(miss).size(), 0);
+    }
+
+    @Test
+    public void testSingletonFlatbushXY()
+    {
+        // Because mixing up x and y is easy to do...
+        List<Rectangle> items = new ArrayList();
+        items.add(new Rectangle(0, 10, 1, 11));
+        Flatbush rtree = new Flatbush<Rectangle>(items.toArray(new Rectangle[] {}));
+        Rectangle hit = new Rectangle(1, 11, 2, 12);
+        assertEquals(rtree.findIntersections(hit), items);
+        Rectangle miss = new Rectangle(11, 1, 12, 2);
+        assertEquals(rtree.findIntersections(miss).size(), 0);
+    }
+
+    @Test
+    public void testDoubletonFlatbush()
+    {
+        // This is the smallest Rtree with height > 1
+        // Also test for some degeneracies
+        List<Rectangle> items = new ArrayList();
```

Comment:
`assertEqualsNoOrder(List<Rectangle>, List<Rectangle>)` helper function might be useful here:

```
@Test
    public void testDoubletonFlatbush()
    {
        // This is the smallest Rtree with height > 1
        // Also test for some degeneracies
        Rectangle rect0 = new Rectangle(1, 1, 1, 1);
        Rectangle rect1 = new Rectangle(-1, -2, -1, -1);

        List<Rectangle> items = ImmutableList.of(rect0, rect0);
        Flatbush<Rectangle> rtree = new Flatbush<>(items.toArray(new Rectangle[] {}));

        assertEqualsNoOrder(findIntersections(rtree, EVERYTHING), items);

        assertEquals(findIntersections(rtree, new Rectangle(1, 1, 2, 2)), ImmutableList.of(rect0));

        assertEquals(findIntersections(rtree, new Rectangle(-2, -2, -1, -2)), ImmutableList.of(rect1));
        // This should test missing at the root level
        assertEquals(findIntersections(rtree, new Rectangle(10, 10, 12, 12)), ImmutableList.of());
        // This should test missing at the leaf level
        Rectangle marginalMiss = new Rectangle(0, 0, 0, 0);
        assertEquals(findIntersections(rtree, new Rectangle(0, 0, 0, 0)), ImmutableList.of());
    }

    private static void assertEqualsNoOrder(List<Rectangle> actual, List<Rectangle> expected)
    {
        List<Rectangle> actualSorted = actual.stream().sorted(RECTANGLE_COMPARATOR).collect(toImmutableList());
        List<Rectangle> expectedSorted = expected.stream().sorted(RECTANGLE_COMPARATOR).collect(toImmutableList());

        assertEquals(actualSorted, expectedSorted);
    }
```